### ISSUE TITLE
create a uniform approach for CSV output

### ISSUE THREAD

**Issue by simsong** [sleuthkit/sleuthkit#3086]
The problem with https://github.com/sleuthkit/sleuthkit/pull/3072 is that it modifies every output line with an `if (csv)` to get csv output. This is tedious and error-prone. A better approach would be to have a new formatted `printf()` that extracts the format statements and outputs in CSV format if a format is set, otherwise uses the full format information.

- [ ] Implement a unified formatter that either outputs in CSV format or using the format string.
- [ ] Adopt https://github.com/sleuthkit/sleuthkit/pull/3072 to use it.
- [ ] Add a test for both formatted output and CSV output for mmls.

**Comment by uckelman-sf** [comment]
A better approach would be to have an object implementing a formatter interface that gets called from the walker functions, so as to extract the formatting out of the walker completely.

**Comment by simsong** [comment]
Yes. That is my goal. Have a new class that does the formatted printing
that takes the format strings, the list of things to be output, and a mode.

**Comment by uckelman-sf** [comment]
I wouldn't pass the class the format strings. Encapsulate them inside the class. Pass in the data only.

**Comment by uckelman-sf** [comment]
I've reread the code in mmls.cpp now and have changed my mind. `part_act` has so little other logic in it that it basically already is a formatter. The simplest thing to read and maintain would be to have two versions of `part_act` (and two of `print_header`) so that the CSV writing code isn't mixed with the standard ouptut.

**Comment by uckelman-sf** [comment]
This
```if (tsk_vs_part_walk(vs.get(), 0, vs->part_count - 1,
            (TSK_VS_PART_FLAG_ENUM) flags, part_act, &ws)) {
        tsk_error_print(stderr);
        return 1;
    }
```
would become this:
```
if (tsk_vs_part_walk(vs.get(), 0, vs->part_count - 1,
            (TSK_VS_PART_FLAG_ENUM) flags, part_act, &ws)) {
        tsk_error_print(stderr);
        return 1;
    }
```if (tsk_vs_part_walk(vs.get(), 0, vs->part_count - 1,
            (TSK_VS_PART_FLAG_ENUM) flags, is_csv ? part_act_csv : part_act, &ws)) {
        tsk_error_print(stderr);
        return 1;
    }
```
No classes needed, just two functions with the same signature.

**Comment by simsong** [comment]
> I've reread the code in mmls.cpp now and have changed my mind. `part_act` has so little other logic in it that it basically already is a formatter. The simplest thing to read and maintain would be to have two versions of `part_act` (and two of `print_header`) so that the CSV writing code isn't mixed with the tabular ouptut.

I don't like all of the `if` statements. I am thinking of a class for which:
* It has a number of registered formatters (like [format statements in FORTRAN](https://web.stanford.edu/class/me200c/tutorial_77/17_format.html))
* An output method that takes a list of things to output and a specific format to use
* An output mode that tells the class to either use the formatter or to output as CSV (with the formatter code optionally as the first column, to aid in parsing)
Then we can trivially add JSON or XML output as well.

**Comment by simsong** [comment]
> The approach I suggested would eliminate all the conditionals. Adding classes here is overengineering. Building something more generic will take days to do something that could be done in about 20 minutes.

Okay.

**Comment by uckelman-sf** [comment]
I've pushed an example to #3090.
